gitlab-runner/src/lib.rs

```diff
@@ -106,6 +106,36 @@ macro_rules! outputln {
 pub type JobResult = Result<(), ()>;
 pub use client::Phase;
 
+/// A file that a [`JobHandler`] is willing to upload to the server
+///
+/// A [`JobHandler`] will be queried for the set of files it is able
+/// to upload to the server. These might be on disk, or they might be
+/// generated or downloaded from some other source on demand. The
+/// `get_path` method is required so that the globbing Gitlab expects
+/// can be performed without the handler needing to be involved.
+pub trait UploadableFile: Eq {
+    /// Get the logical path of the file.
+    ///
+    /// This is the path on disk from the root of the checkout for
+    /// Gitlab's own runner. It should match the paths that users are
+    /// expected to specify when requesting artifacts in their jobs.
+    fn get_path(&self) -> String;
```

Comment:
why `String` instead of `&str`

Reply:
I did it this way because for the runner I was working with, it's convenient to generate the path names on demand (as e.g. `format!("{}_log.yaml", self.job_id)`) and making it a `&str` gives us headaches there (it effectively forces us to store the string somewhere, and there's no particular reason to). 

In general, I assume for non-standard runners, it's quite likely that you aren't really reading files off the filesystem, you're handing out promises to download or generate something from somewhere else, if the path you register it under is requested by the job. In that case, it's convenient to generate the filename on demand.

However, if there's an efficiency concern about repeatedly generating the filenames here, I could easily force the client crate to store them?

Reply:
Yeah the user could keep a cache of UploadFiles in that case indeed with an allocated string. Though likely it's not going to be a real concern

Reply:
If you want to easily support both cases you could make it Cow<str>  as well ofcourse